core/trino-main/src/main/java/io/trino/metadata/Metadata.java

```diff
@@ -732,6 +732,11 @@ default boolean isMaterializedView(Session session, QualifiedObjectName viewName
      */
     boolean supportsReportingWrittenBytes(Session session, QualifiedObjectName tableName, Map<String, Object> tableProperties);
 
+    /**
+     * Returns true if a connector supports multiple writers to write the same partition concurrently for a partitioned table. Otherwise, it returns false.
+     */
+    boolean supportsMultipleWritersPerPartition(Session session, QualifiedObjectName schemaTableName);
```

Comment:
I had a similar dilemma when working on https://github.com/trinodb/trino/pull/14735.

The way I decided to solve it for now is to follow the `preferred` partitioning semantic. Basically if the explicit partitioning requirement is set in https://github.com/trinodb/trino/blob/master/core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorTableLayout.java#L23 then it indicates that partitioning cannot be violated (must be exact). However when the partitioning is set based on the `ConnectorTableLayout#partitionColumn` it indicates that data is grouped in a "preferred" way set by the engine and the partitioning doesn't have to be followed exactly. I'm happy to discuss what do we think is the right approach so we can consolidate. 

CC: @sopel39 @gaurav8297 @martint

Reply:
>  I'm happy to discuss what do we think is the right approach so we can consolidate.

I think method like here should be sufficient. Wdyt?

> However when the partitioning is set based on the ConnectorTableLayout#partitionColumn it indicates that data is grouped in a "preferred" way that is set by the engine and the partitioning requirement doesn't have to be followed exactly allowing potential optimizations

Preferred never meant 1-1 mapping. Comment in `io.trino.spi.connector.ConnectorTableLayout#ConnectorTableLayout(java.util.List<java.lang.String>)`:
```
     * Creates a preferred table layout that is evenly partitioned on given columns by the engine.
     * Such layout might be ignored by Trino planner.
```
Engine is free to ignore it and split partition data between many writers

Reply:
> I think method like here should be sufficient. Wdyt?

I wonder if this method is necessary. I was trying to suggest that instead we can rely on the `ConnectorTableLayout`. For example if the `ConnectorTableLayout` has a specific partitioning handle set it indicates "strict" partitioning. However if the partitioning handle is not set, and the partitioning is configured based on `partitionColumns` it indicates exactly what you say, `Preferred never meant 1-1 mapping`, allowing the engine to violate partition assignment and round robin data from the same partition between multiple writers.

Reply:
> I wonder if this method is necessary. I was trying to suggest that instead we can rely on the ConnectorTableLayout. For example if the ConnectorTableLayout has a specific partitioning handle set it indicates "strict" partitioning.

We want to support scaling even for partitioning which seem strict. E.g. Currently, engine considers Hive bucketing to be "strict" (e.g. one writer per parition). However, it's not really necessary and engine can use more writers per bucket.

Check implementations of `supportsMultipleWritersPerPartition` in Hive, Delta, Iceberg. Previously scaling of such inserts didn't happen because of "strict" partitioning

Reply:
> However, it's not really necessary and engine can use more writers per bucket.

Then a connector doesn't have to set a strict `partitioningHandle` in the `ConnectorTableLayout` and only provide a list of "partitionColumns" that also include bucket columns. Or am I missing something?

Reply:
As per our discussion, I've added a flag inside `ConnectorTableLayout` instead of `supportsMultipleWritersPerPartition` method inside `Metadata` cc @arhimondr @sopel39